### PR TITLE
Configure Github CI to run tests on Windows

### PR DESCRIPTION
**Pull request by Sgitario**
Changes:
- Windows fails to stop an application when running on DEV mode. Reported issue: https://github.com/quarkusio/quarkus/issues/14647
- Windows disallows to delete folders when they are being used by the file explorer or the processes are about to be exited. This is why:
1. I'm ignoring these errors when the directory can't be deleted (it's not Quarkus issue)
2. I splitted the "s p a c e s" runs into there different folders for jvm, native and dev mode.

### CODE REVIEW

**Comment by Sgitario** [comment]
1.11.X is not working on Native in Windows due to https://github.com/quarkusio/quarkus/issues/7990. Therefore, we need to upgrade to 1.13.X to make Windows work on Native: https://github.com/quarkus-qe/quarkus-startstop/pull/76

File: testsuite/src/it/java/io/quarkus/ts/startstop/SpecialCharsTest.java [hunk 1]
```diff
@@ -155,28 +152,24 @@ public void testRuntime(TestInfo testInfo, Apps app, MvnCmds mvnCmds, String sub
...
-            }
+            removeDirWithSpecialCharacters(appDestDir);
         }
     }
 
-
     @Test
     public void spacesJVM(TestInfo testInfo) throws IOException, InterruptedException {
-        testRuntime(testInfo, Apps.JAX_RS_MINIMAL, MvnCmds.JVM, "s p a c e s");
+        testRuntime(testInfo, Apps.JAX_RS_MINIMAL, MvnCmds.JVM, "s p a c e s j v m");
```
Comment (Sgitario) [inline testsuite/src/it/java/io/quarkus/ts/startstop/SpecialCharsTest.java:161]: Explanation about this change in PR description

File: testsuite/src/it/java/io/quarkus/ts/startstop/SpecialCharsTest.java [hunk 2]
```diff
@@ -155,28 +152,24 @@ public void testRuntime(TestInfo testInfo, Apps app, MvnCmds mvnCmds, String sub
             }
             writeReport(cn, mn, whatIDidReport.toString());
 
-            // Remove dir with special chars
-            if (appDestDir.exists()) {
-                FileUtils.deleteDirectory(appDestDir);
-            }
+            removeDirWithSpecialCharacters(appDestDir);
```
Comment (Sgitario) [inline testsuite/src/it/java/io/quarkus/ts/startstop/SpecialCharsTest.java:155]: Explanation about this change in PR description

File: testsuite/src/it/java/io/quarkus/ts/startstop/SpecialCharsTest.java [hunk 3]
```diff
@@ -155,28 +152,24 @@ public void testRuntime(TestInfo testInfo, Apps app, MvnCmds mvnCmds, String sub
...
     @Test
     public void spacesJVM(TestInfo testInfo) throws IOException, InterruptedException {
-        testRuntime(testInfo, Apps.JAX_RS_MINIMAL, MvnCmds.JVM, "s p a c e s");
+        testRuntime(testInfo, Apps.JAX_RS_MINIMAL, MvnCmds.JVM, "s p a c e s j v m");
     }
 
     @Test
     public void spacesDEV(TestInfo testInfo) throws IOException, InterruptedException {
-        testRuntime(testInfo, Apps.JAX_RS_MINIMAL, MvnCmds.DEV, "s p a c e s");
+        testRuntime(testInfo, Apps.JAX_RS_MINIMAL, MvnCmds.DEV, "s p a c e s d e v");
```
Comment (Sgitario) [inline testsuite/src/it/java/io/quarkus/ts/startstop/SpecialCharsTest.java:166]: Explanation about this change in PR description

File: testsuite/src/it/java/io/quarkus/ts/startstop/SpecialCharsTest.java [hunk 4]
```diff
@@ -155,28 +152,24 @@ public void testRuntime(TestInfo testInfo, Apps app, MvnCmds mvnCmds, String sub
...
     }
 
     @Test
     @Tag("native")
     public void spacesNative(TestInfo testInfo) throws IOException, InterruptedException {
-        testRuntime(testInfo, Apps.JAX_RS_MINIMAL, MvnCmds.NATIVE, "s p a c e s");
+        testRuntime(testInfo, Apps.JAX_RS_MINIMAL, MvnCmds.NATIVE, "s p a c e s n a t i v e");
```
Comment (Sgitario) [inline testsuite/src/it/java/io/quarkus/ts/startstop/SpecialCharsTest.java:172]: Explanation about this change in PR description

File: testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java [hunk 5]
```diff
@@ -77,6 +82,25 @@
             // (no explicit configuration, none or multiple JDBC driver extensions)
             // Result of DevServices support https://github.com/quarkusio/quarkus/pull/14960
             Pattern.compile(".*Unable to determine a database type for default datasource.*"),
+    }),
+    WINDOWS_COMMON(new Pattern[] {
+            // When running on Windows, some warning SLF4J traces are randomly printed
+            Pattern.compile(".*SLF4J: .*")
```
Comment (Sgitario) [inline testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java:88]: I didn't report this issue as I can't reproduce it locally and it's only printed randomly and as warnings, so it's not a big issue.

Comment (rsvoboda) [reply testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java:88]: This is weird, just on Windows? We had issues with leaking of this warning on Linux some time ago too.

We have explicit check for sl4j - https://github.com/quarkus-qe/quarkus-startstop/pull/45/files

File: testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java [hunk 6]
```diff
@@ -77,6 +82,25 @@
...
+    }),
+    WINDOWS_DEV_MODE_ERRORS(new Pattern[]{
+            Pattern.compile(".*Re-run Maven using the -X switch to enable full debug logging.*"),
```
Comment (rsvoboda) [inline testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java:91]: Add hint what is causing these errors

File: testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java [hunk 7]
```diff
@@ -77,6 +82,25 @@
...
+            Pattern.compile(".*For more information about the errors and possible solutions, please read the following articles.*"),
+            Pattern.compile(".*Failed to run: Dev mode process did not complete successfully.*"),
+            Pattern.compile(".*http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException.*"),
+            Pattern.compile(".*To see the full stack trace of the errors, re-run Maven with the -e switch.*"),
+            Pattern.compile("\\[ERROR\\] *"),
+    }),
+    WINDOWS_NATIVE(new Pattern[]{
+            Pattern.compile(".*Uber JAR strategy is used for native image source JAR generation on Windows.*"),
```
Comment (rsvoboda) [inline testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java:99]: Why is there this special handling of Windows?

File: testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java [hunk 8]
```diff
@@ -77,6 +82,25 @@
...
+            Pattern.compile(".*Duplicate entry about.html entry.*"),
```
Comment (rsvoboda) [inline testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java:100]: Why is this Windows specific?

File: testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java [hunk 9]
```diff
@@ -77,6 +82,25 @@
...
+            Pattern.compile(".*Dependencies with duplicate files detected.*"),
+            Pattern.compile(".*objcopy executable not found in PATH. Debug symbols will not be separated from executable.*"),
```
Comment (rsvoboda) [inline testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java:102]: We have the same issue handled a bit differently - https://github.com/quarkus-qe/quarkus-startstop/blob/main/testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java#L88

TBH I'm not sure if we need to handle special WINDOWS_COMMON / WINDOWS_DEV_MODE_ERRORS / WINDOWS_NATIVE.
I would just go with the `platformErrs` approach, merge ^^^ into one array

**Review by rsvoboda — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yaml [hunk 10]
```diff
@@ -96,4 +96,95 @@ jobs:
...
+          gu.cmd install native-image
+      - name: Configure Pagefile
+        # Increased the page-file size due to memory-consumption of native-image command
+        # For details see https://github.com/actions/virtual-environments/issues/785
+        uses: al-cheb/configure-pagefile-action@v1.2
+      # We do this so we can get better analytics for the downloaded version of the build images
+      - name: Update Docker Client User Agent
+        shell: bash
+        run: |
+          cat <<< $(jq '.HttpHeaders += {"User-Agent": "Quarkus-CI-Docker-Client"}' ~/.docker/config.json) > ~/.docker/config.json
```
[the remark this answers is not in the document]
Comment (Sgitario) [reply .github/workflows/ci.yaml:174]: This is a copy of what Quarkus upstream is using, we can get rid of this.

[the remark this answers is not in the document]
Comment (Sgitario) [reply .github/workflows/ci.yaml:174]: Deleted

**Comment by Sgitario** [comment]
PR updated with the following changes:
- Move Windows specified whitelisted errors into the existing platformErrs approach
- Not to ignore SLF4J errors. If this happens again with Windows, it will fail.

I will mark all the comments as resolved once I'm happy with the new changes and the CI passes.

File: testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java (hunk 6 above)
Comment (Sgitario) [reply testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java:84]: Added

File: testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java (hunk 8 above)
Comment (Sgitario) [reply testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java]: Moved to common platform whitelist log lines.

File: testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java (hunk 7 above)
Comment (Sgitario) [reply testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java]: Moved to Windows platform whitelist log lines.

**Comment by Sgitario** [comment]
@rsvoboda would you mind to take another look? I moved the windows errors to platform windows errors, but kept the ones about dev console as this is an issue that should be removed once solved.
About the slf4j errors, this is not whitelisted any longer, so if it happens again, the test will fail as expected.

File: testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java [hunk 11]
```diff
@@ -77,7 +77,24 @@
...
+            Pattern.compile(".*For more information about the errors and possible solutions, please read the following articles.*"),
+            Pattern.compile(".*Failed to run: Dev mode process did not complete successfully.*"),
+            Pattern.compile(".*http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException.*"),
+            Pattern.compile(".*To see the full stack trace of the errors, re-run Maven with the -e switch.*"),
+            Pattern.compile("\\[ERROR\\] *"),
     });
+	
+	// Randomly fails when vertx-cache temporary directory exists. Related to https://github.com/quarkusio/quarkus/issues/7678
+    // And https://github.com/quarkusio/quarkus/pull/15541/files#diff-a38e0d86cf6a637c19b6e0a0e23959f644886bdcc0f0e5615ce7cfa0e6bc9909R244
+	private static final Pattern COMMON_WARNING_VERTX_CACHE_DIRECTORY = Pattern.compile(".*Unable to make the Vert.x cache directory.*");
```
Comment (rsvoboda) [inline testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java:94]: Are these 3 really happening on all the platforms or just Windows?

File: testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java [hunk 12]
```diff
@@ -89,9 +106,28 @@
         switch (OS.current()) {
             case MAC:
                 return new Pattern[] {
+                		COMMON_WARNING_VERTX_CACHE_DIRECTORY,
+                		COMMON_WARNING_DUPLICATE_ENTRY_NATIVE,
+                		COMMON_WARNING_DUPLICATE_ENTRIES_NATIVE,
+                        Pattern.compile(".*objcopy executable not found in PATH. Debug symbols will not be separated from executable.*"),
```
Comment (rsvoboda) [inline testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java:112]: should be constants, shared on Win + Mac

**Review by rsvoboda — CHANGES_REQUESTED** [review]

File: testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java (hunk 11 above)
Comment (Sgitario) [reply testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java:94]: As I could see in the PR, it could happen in any platform.

Comment (rsvoboda) [reply testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java]: And does it happen? TBH it should not happen at all as a random directory is generated every time.

File: testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java (hunk 12 above)
Comment (Sgitario) [reply testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java]: Done

**Comment by Sgitario** [comment]
PR updated to move some shared constants between Windows and Mac.

File: testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java (hunk 11 above)
Comment (Sgitario) [reply testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java]: It happened to me when trying on Windows. However, I will remove it and let's see.

**Comment by Sgitario** [comment]
I've updated the PR to whitelist the "unable to create directory for vert.x cache" error only for Windows. This is reported by https://github.com/quarkusio/quarkus/issues/16895.

File: testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java (hunk 5 above)
Comment (Sgitario) [reply testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java]: I reverted these changes.

File: testsuite/src/it/java/io/quarkus/ts/startstop/utils/Logs.java [hunk 13]
```diff
@@ -94,6 +95,24 @@ public static void checkLog(String testClass, String testMethod, Apps app, MvnCm
                     }
                 }
             }
+
+            List<Pattern> ignoredLines = new ArrayList<>(Arrays.asList(WhitelistLogLines.COMMON.errs));
```
[the remark this answers is not in the document]
Comment (Sgitario) [reply testsuite/src/it/java/io/quarkus/ts/startstop/utils/Logs.java]: Changed.

File: testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java (hunk 11 above)
Comment (Sgitario) [reply testsuite/src/it/java/io/quarkus/ts/startstop/utils/WhitelistLogLines.java]: Changed to be only for Windows and reported.

**Review by rsvoboda — APPROVED** [review]

**Pull request merged**